addons/crm_mail_plugin/controllers/mail_plugin.py

```diff
@@ -1,81 +1,153 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import logging
+from werkzeug.exceptions import Forbidden
 
-from odoo.http import request
+from odoo.fields import Domain
+from odoo.http import request, route
+from odoo.tools import html2plaintext, email_normalize
 from odoo.tools.misc import formatLang
+from odoo.tools.image import image_data_uri
 
 from odoo.addons.mail_plugin.controllers import mail_plugin
 
 _logger = logging.getLogger(__name__)
 
 
 class MailPluginController(mail_plugin.MailPluginController):
+    def _get_record_redirect_url(self, model, record_id):
+        if model == 'crm.lead':
+            return f'/odoo/crm/{int(record_id)}'
 
-    def _fetch_partner_leads(self, partner, limit=5, offset=0):
-        """
-        Returns an array containing partner leads, each lead will have the following structure :
-        {
-            id: the lead's id,
-            name: the lead's name,
-            expected_revenue: the expected revenue field value
-            probability: the value of the probability field,
-            recurring_revenue: the value of the recurring_revenue field if the lead has a recurring revenue
-            recurring_plan: the value of the recurring plan field if the lead has a recurring revenue
-        }
-        """
+        return super()._get_record_redirect_url(model, record_id)
 
-        partner_leads = request.env['crm.lead'].search(
-            [('partner_id', '=', partner.id)], offset=offset, limit=limit)
-        recurring_revenues = request.env.user.has_group('crm.group_use_recurring_revenues')
+    @route()
+    def search_records(self, model, query, limit=30):
+        if model == "crm.lead":
+            if model not in self._mail_models_access_whitelist('read'):
```

Comment:
Feels like this should be a generic thing in the main controller (aka call super that does checks then return stuff)